distributed/protocol/numpy.py

```diff
@@ -105,33 +105,41 @@ def deserialize_numpy_ndarray(header, frames):
 
 @dask_serialize.register(np.ma.core.MaskedArray)
 def serialize_numpy_maskedarray(x):
-    # Separate elements of the masked array that we need to deal with discretely.
-    data = x.data
-    mask = x.mask
-    fill_value = x.fill_value
-
-    # Make use of existing numpy serialization for the two ndarray elements of
-    # the masked array.
-    data_header, data_frames = serialize_numpy_ndarray(data)
-    mask_header, mask_frames = serialize_numpy_ndarray(mask)
+    data_header, frames = serialize_numpy_ndarray(x.data)
+    header = {'data-header': data_header,
+              'nframes': len(frames)}
+
+    # Serialize mask if present
+    if x.mask is not np.ma.nomask:
+        mask_header, mask_frames = serialize_numpy_ndarray(x.mask)
+        header['mask-header'] = mask_header
+        frames += mask_frames
+
+    # Only a few dtypes have python equivalents msgpack can serialize
+    if isinstance(x.fill_value, (np.integer, np.floating, np.bool_)):
+        serialized_fill_value = (False, x.fill_value.item())
+    else:
+        serialized_fill_value = (True, pickle.dumps(x.fill_value))
```

Comment:
I suggest that we just always pickle for simplicity.  I don't think that this will affect performance much.

Reply:
Perhaps, although pickling a simple numpy scalar results in an unfortunate number of bytes:

```
In [12]: len(msgpack.dumps(pickle.dumps(x)))
Out[12]: 120

In [13]: len(msgpack.dumps(x.item()))
Out[13]: 9

In [14]: %timeit msgpack.dumps(pickle.dumps(x))
10.3 µs ± 157 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [15]: %timeit msgpack.dumps(x.item())
1.72 µs ± 22.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
Hrm, that is surprisingly expensive.  OK, seems reasonable to me.